pkg/kv/kvserver/apply/asserter.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+// TODO: put in a test package
+package apply
+
+import (
+	fmt "fmt"
+
+	"github.com/cockroachdb/cockroach/pkg/kv/kvpb"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvserverbase"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"go.etcd.io/raft/v3/raftpb"
+)
+
+type appliedCmd struct {
+	cmdID    kvserverbase.CmdIDKey
+	rangeID  roachpb.RangeID
+	logIndex kvpb.RaftIndex
+	replicas map[roachpb.ReplicaID]bool
+}
+
+// Asserter is a test utility that tracks application of Raft commands, and
+// asserts that a command is only applied once per replica, and never applied at
+// different log indexes (i.e. a double-apply or replay).
+//
+// It accounts for snapshots, but not splits/merges which are considered
+// separate Raft groups, as the risk of proposals leaking across Raft groups
+// appears negligible.
+type Asserter struct {
+	mu syncutil.Mutex
+	// appliedIndex tracks the applied index of each replica.
+	appliedIndex map[roachpb.RangeID]map[roachpb.ReplicaID]kvpb.RaftIndex
+	// appliedCmds tracks applied commands by range and command ID.
+	appliedCmds map[roachpb.RangeID]map[kvserverbase.CmdIDKey]*appliedCmd
+}
+
+// NewAsserter creates a new asserter.
+func NewAsserter() *Asserter {
+	return &Asserter{
+		appliedIndex: map[roachpb.RangeID]map[roachpb.ReplicaID]kvpb.RaftIndex{},
+		appliedCmds:  map[roachpb.RangeID]map[kvserverbase.CmdIDKey]*appliedCmd{},
+	}
+}
+
+// ensureRangeLocked ensures that data structures exist for the given range.
+// Asserter.mu must be held.
+func (a *Asserter) ensureRangeLocked(rangeID roachpb.RangeID) {
+	if _, ok := a.appliedCmds[rangeID]; !ok {
+		a.appliedCmds[rangeID] = map[kvserverbase.CmdIDKey]*appliedCmd{}
+	}
+	if _, ok := a.appliedIndex[rangeID]; !ok {
+		a.appliedIndex[rangeID] = map[roachpb.ReplicaID]kvpb.RaftIndex{}
+	}
+}
+
+// Apply records a command being applied, and asserts validity.
+//
+// It assumes that entry application is durable, i.e. that nodes won't lose
+// applied state. This is not true for real CRDB nodes which can lose applied
+// state on restart, but is generally true in tests. If needed, we can loosen
```

Comment:
This wouldn't be true for tests that restart nodes, so I'd drop "generally true in tests".

Can more clearly say that it's true while the CRDB process runs. And when it restarts, the state can be rolled back.